docs/concepts/The_Context_Type.md

```diff
@@ -0,0 +1,311 @@
+tags: SmokeFramework, context type
+---
+# The Context Type
+
+The Context Type is a foundational concept of the smoke-framework. The Context Type is designed to provide access to all the *things* an operation handler
+may want to access during that operation's execution. These *things* may include-
+
+1. Service Clients
+2. Helper objects such as loggers
+3. Injected logic that is dependent on the higher level environment
+
+By default, the instances of the Context Type will be request-scoped - it will be created when a request is received by the framework and will live until 
+the request is fully processed. The Context Type is thus named because it represents the *context* in which the request is being processed by the 
+operation handler.
+
+## Defining the Request Type
+
+Applications initialized using `smoke-framework-application-generate` will be created with a stub definition
+of the Context Type-
+
+```
+public struct MyServiceOperationsContext {
+    let logger: Logger
+    // TODO: Add properties to be accessed by the operation handlers
+    
+    public init(logger: Logger) {
+        self.logger = logger
+    }
+}
+```
+
+The following sections will discuss how to add particular types of instances to the Context Type.
+
+### Helper objects
+
+Helper objects such as loggers are pretty straight forward to add to the context type as typically it is always the same concrete type.
+The stub Context Type created when using `smoke-framework-application-generate` already adds the logger. Other helper objects can be added
+in the same way.
+
+```
+public struct MyServiceOperationsContext {
+    let logger: Logger
+    let myHelperObject: MyHelperObject
+    
+    public init(logger: Logger,
+                myHelperObject: MyHelperObject) {
+        self.logger = logger
+        self.myHelperObject = myHelperObject
+    }
+}
+```
+
+In your application handler you can then make use of this helper instance-
+
+```
+extension MyServiceOperationsContext {
+    public func handleHasSubscriptions(input: MyServiceModel.HasSubscriptionsRequest) async throws
+    -> MyServiceModel.HasSubscriptionsResponse {
+        self.myHelperObject.doTheThing()
+                
+        return HasSubscriptionsResponse(hasSubscriptions: false)
+    }
+}
+```
+
+You will then have to update the application's initializer to provide an instance appropriate to each request.
+
+For unit tests, you will also need to provide an instance appropriate to each unit test.
+
+```
+func createOperationsContext(myHelperObject myHelperObjectOptional: MyHelperObject? = nil) 
+-> MyServiceOperationsContext {
+    let myHelperObject = myHelperObjectOptional ?? getDefaultMyHelperObjectForTesting()
+    
+    return MyServiceOperationsContext(logger: TestVariables.logger,
+                                      myHelperObject: myHelperObject)
+}
+```
+
+### Service Clients
```

Comment:
I'd like to see some more context about the performance difference of the protocol approach vs the generic approach for service clients. As I understand, existential box redirection is analogous to v-table lookup, which should have a very small perf hit, especially when considering that the underlying operation of the service clients will be a network call.

The generic approach seems much more complex code-wise, especially where there are more than just a couple of service clients, so having more definitive information about why one would go with one approach over the other would be helpful.